src/modules/luksbootkeyfile/LuksBootKeyFileJob.cpp

```diff
@@ -137,32 +137,58 @@ generateTargetKeyfile()
 {
-    // Sometimes this error is thrown: "All key slots full"
-    // luksAddKey will fail. So, remove the first slot to make room
+    // Get luksDump for this device
+    int slots_count = 0;
     auto luks_dump = CalamaresUtils::System::instance()->targetEnvCommand(
-        { "cryptsetup", "luksDump", d.device }, QString(), QString(), std::chrono::seconds( 5 ) );
-    if ( luks_dump.getExitCode() == 0 )
+        { QStringLiteral( "cryptsetup" ), QStringLiteral( "luksDump" ), d.device },
+        QString(),
+        QString(),
+        std::chrono::seconds( 5 ) );
+    if ( luks_dump.getExitCode() != 0 )
     {
-        QRegularExpression re( QStringLiteral( R"(\d+:\s*enabled)" ), QRegularExpression::CaseInsensitiveOption );
-        int count = luks_dump.getOutput().count( re );
-        cDebug() << "Luks Dump slot count: " << count;
-        if ( count >= 7 )
-        {
-            auto r = CalamaresUtils::System::instance()->targetEnvCommand(
-                { "cryptsetup", "luksKillSlot", d.device, "1" }, QString(), d.passphrase, std::chrono::seconds( 60 ) );
-            if ( r.getExitCode() != 0 )
-            {
-                cWarning() << "Could not kill a slot to make room on" << d.device << ':' << r.getOutput()
-                           << "(exit code" << r.getExitCode() << ')';
-                return false;
-            }
-        }
+        cWarning() << "Could not get LUKS information on " << d.device << ':' << luks_dump.getOutput() << "(exit code"
+                   << luks_dump.getExitCode() << ')';
+        return false;
     }
 
-    // Adding the key can take some times, measured around 15 seconds with
-    // a HDD (spinning rust) and a slow-ish computer. Give it a minute.
+    // Check for LUKS2
+    QString luks2_hash = QString();
+    QRegularExpression pbkdf_re( QStringLiteral( R"(pbkdf:\s*(.*))" ), QRegularExpression::CaseInsensitiveOption );
+    QRegularExpressionMatch match = pbkdf_re.match( luks_dump.getOutput() );
+    if ( match.hasMatch() ) {
+        luks2_hash = match.captured(1);
+        cDebug() << "Setup LUKS2 " << luks2_hash << " for " << d.device;
+    }
+    else
+    {
+        cDebug() << "Setup LUKS1 for " << d.device;
+    }
+
+    // Check the number of slots used
+    // Output of LUKS1 and LUKS2 differ
+    auto search_pattern = luks2_hash.isEmpty() ? QStringLiteral( R"(\d+:\s*enabled)" ) : QStringLiteral( R"(\d+:\s*luks2)" );
+    QRegularExpression slots_re( search_pattern, QRegularExpression::CaseInsensitiveOption );
+    slots_count = luks_dump.getOutput().count( slots_re );
+
+    if ( ( luks2_hash.isEmpty() && slots_count == 8 ) ||
+         ( !luks2_hash.isEmpty() && slots_count == 32 ))
+    {
+        // No key slots left: return gracefully
+        return true;
+    }
+
+    // Add the key to the keyfile
+    QStringList args = { QStringLiteral( "cryptsetup" ), QStringLiteral( "luksAddKey" ) };
+    if ( !luks2_hash.isEmpty() )
+    {
+        args << "--pbkdf" << luks2_hash;
```

Comment:
I am not sure that grabbing the algorithm from another key is the right thing here.  I would say that we should either use the default or add a config option for the user to specify it.

Reply:
The LUKS2 hash (pbkdf, argon2i, argon2id) should be configured in the partition module. It is needed when creating the encrypted partitions. Here, we add a key to the keyfile and for LUKS2 that means we need to provide the hash key used when creating the encrypted partition. At least, that's what I understood and that's why I checked for the used hash of that partition. Am I wrong?

Reply:
You could look at the code in the partition module to verify but IIRC you can't select it there.  It is using the default.

Either way, I think it is better to either use the default here or offer a config option.